src/System.Security.Cryptography.Pkcs/tests/EnvelopedCms/StateTests.cs

```diff
@@ -0,0 +1,357 @@
+        }
+
+        [Fact]
+        public static void PostEncrypt_RecipientInfos()
+        {
+            ContentInfo expectedContentInfo = new ContentInfo(new byte[] { 1, 2, 3 });
+            EnvelopedCms ecms = new EnvelopedCms(expectedContentInfo);
+            using (X509Certificate2 cert = Certificates.RSAKeyTransfer1.GetCertificate())
+            {
+                ecms.Encrypt(new CmsRecipient(cert));
+            }
+
+            object ignore;
+            Assert.ThrowsAny<CryptographicException>(() => ignore = ecms.RecipientInfos);
+        }
+
+        [Fact]
+        public static void PostEncrypt_Decrypt()
+        {
+            ContentInfo expectedContentInfo = new ContentInfo(new byte[] { 1, 2, 3 });
+            EnvelopedCms ecms = new EnvelopedCms(expectedContentInfo);
+            using (X509Certificate2 cert = Certificates.RSAKeyTransfer1.GetCertificate())
+            {
+                ecms.Encrypt(new CmsRecipient(cert));
+            }
+            Assert.ThrowsAny<CryptographicException>(() => ecms.Decrypt());
+        }
+
+        [Fact]
+        public static void PostEncrypt_ContentInfo()
+        {
+            ContentInfo expectedContentInfo = new ContentInfo(new byte[] { 1, 2, 3 });
+            EnvelopedCms ecms = new EnvelopedCms(expectedContentInfo);
+            using (X509Certificate2 cert = Certificates.RSAKeyTransfer1.GetCertificate())
+            {
+                ecms.Encrypt(new CmsRecipient(cert));
+            }
+
+            // Encrypting does not update ContentInfo.
+            ContentInfo actualContentInfo = ecms.ContentInfo;
+            Assert.Equal(expectedContentInfo.ContentType, actualContentInfo.ContentType);
+            Assert.Equal<byte>(expectedContentInfo.Content, actualContentInfo.Content);
+        }
+
+        //
+        // State 3: Called Decode()
+        //
+
+        [Fact]
+        public static void PostDecode_Encode()
+        {
+            byte[] encodedMessage =
+                ("3082010c06092a864886f70d010703a081fe3081fb0201003181c83081c5020100302e301a311830160603550403130f5253"
+                + "414b65795472616e7366657231021031d935fb63e8cfab48a0bf7b397b67c0300d06092a864886f70d01010105000481805e"
+                + "bb2d08773594be9ec5d30c0707cf339f2b982a4f0797b74d520a0c973d668a9a6ad9d28066ef36e5b5620fef67f4d79ee50c"
+                + "25eb999f0c656548347d5676ac4b779f8fce2b87e6388fbe483bb0fcf78ab1f1ff29169600401fded7b2803a0bf96cc160c4"
+                + "96726216e986869eed578bda652855c85604a056201538ee56b6c4302b06092a864886f70d010701301406082a864886f70d"
+                + "030704083adadf63cd297a86800835edc437e31d0b70").HexToByteArray();
+
+            EnvelopedCms ecms = new EnvelopedCms();
+            ecms.Decode(encodedMessage);
+
+            // This should really have thrown an InvalidOperationException. Instead, you get... something back.
+            byte[] expectedGarbage = "35edc437e31d0b70".HexToByteArray();
```

Comment:
On Windows, you use the same msgparam type (CMSG_INNER_CONTENT_TYPE_PARAM) to get both the encoding and inner-content – and depending in which stage the cryptmessage handle is in, you get back either the decrypted content, the encrypted content or the wire-format of the CMS. That’s why these two properties often return the same result.

On the desktop, you have the further amusement that the “encrypted inner content” case gets mutated by the “forgot to call Array.Resize” bug.